crates/optimism/trie/src/proof.rs

```diff
@@ -0,0 +1,572 @@
+//! Provides proof operation implementations for [`OpProofsStorage`].
+
+use alloy_primitives::{
+    keccak256,
+    map::{B256Map, HashMap},
+    Address, Bytes, B256, U256,
+};
+use reth_db_api::DatabaseError;
+use reth_execution_errors::{StateProofError, StateRootError, StorageRootError, TrieWitnessError};
+use reth_primitives_traits::Account;
+use reth_trie::{
+    hashed_cursor::{
+        HashedCursor, HashedCursorFactory, HashedPostStateCursorFactory, HashedStorageCursor,
+    },
+    metrics::TrieRootMetrics,
+    proof::{Proof, StorageProof},
+    trie_cursor::{InMemoryTrieCursorFactory, TrieCursor, TrieCursorFactory},
+    updates::TrieUpdates,
+    witness::TrieWitness,
+    AccountProof, BranchNodeCompact, HashedPostState, HashedPostStateSorted, HashedStorage,
+    MultiProof, MultiProofTargets, Nibbles, StateRoot, StorageMultiProof, StorageRoot, TrieInput,
+};
+
+use crate::api::{
+    OpProofsHashedCursor, OpProofsStorage, OpProofsStorageError,
+    OpProofsTrieCursor as OpProofsDBTrieCursor,
+};
+
+/// Manages reading storage or account trie nodes from [`OpProofsDBTrieCursor`].
+#[derive(Debug, Clone)]
+pub struct OpProofsTrieCursor<C: OpProofsDBTrieCursor>(pub C);
+
+impl<C: OpProofsDBTrieCursor> OpProofsTrieCursor<C> {
+    /// Creates a new `OpProofsTrieCursor` instance.
+    pub const fn new(cursor: C) -> Self {
+        Self(cursor)
+    }
+}
+
+impl From<OpProofsStorageError> for DatabaseError {
+    fn from(error: OpProofsStorageError) -> Self {
+        Self::Other(error.to_string())
+    }
+}
+
+impl<C: OpProofsDBTrieCursor + Send + Sync> TrieCursor for OpProofsTrieCursor<C> {
+    fn seek_exact(
+        &mut self,
+        key: Nibbles,
+    ) -> Result<Option<(Nibbles, BranchNodeCompact)>, DatabaseError> {
+        self.0.seek_exact(key).map_err(Into::into)
+    }
+
+    fn seek(
+        &mut self,
+        key: Nibbles,
+    ) -> Result<Option<(Nibbles, BranchNodeCompact)>, DatabaseError> {
+        self.0.seek(key).map_err(Into::into)
+    }
+
+    fn next(&mut self) -> Result<Option<(Nibbles, BranchNodeCompact)>, DatabaseError> {
+        self.0.next().map_err(Into::into)
+    }
+
+    fn current(&mut self) -> Result<Option<Nibbles>, DatabaseError> {
+        self.0.current().map_err(Into::into)
+    }
+}
+
+/// Factory for creating trie cursors for [`OpProofsStorage`].
+#[derive(Debug, Clone)]
+pub struct OpProofsTrieCursorFactory<Storage: OpProofsStorage> {
+    storage: Storage,
+    block_number: u64,
+}
+
+impl<Storage: OpProofsStorage> OpProofsTrieCursorFactory<Storage> {
+    /// Creates a new `OpProofsTrieCursorFactory` instance.
```

Comment:
```suggestion
    /// Creates a new instance.
```
hack is to never mention type or field names in docs unless doc linking since it's prone to doc bugs if the type/field name is changed and manual update of docs is overlooked and compiler will only warn if the doc is linked